### PR TITLE
Fixed issue regarding IndexOutOfBounds when no addresses available

### PR DESCRIPTION
**Pull request by vmaraloiu**
Added fix by throwing an IllegalArgumentException when no addresses are available for binding: 
```
GATEWAY_OPTS=-Djava.net.preferIPv4Stack=true
```
and service accept host set to an IPv6 address.

In progress: adding a test case to validate the behavior.

### CODE REVIEW

File: resource.address/spi/src/main/java/org/kaazing/gateway/resource/address/ResourceAddressFactorySpi.java [hunk 1]
```diff
@@ -148,6 +148,10 @@ public final T newResourceAddress(URI location,
             alternate = address;
         }
 
+        if (addresses.size() == 0) {
```
Comment (jitsni) [inline resource.address/spi/src/main/java/org/kaazing/gateway/resource/address/ResourceAddressFactorySpi.java]: Use addresses.isEmpty()

**Comment by sbadugu** [comment]
Vlad is going to add the test and you can wait until jhe finishes the tests.

**Comment by vmaraloiu** [comment]
I have just added the first test. This is still in progress and should not be merged yet.

**Comment by vmaraloiu** [comment]
@dpwspoon: Hello David, can you please provide some feedback on the current PR status?

File: server/src/test/java/org/kaazing/gateway/server/preferedipv4/ResourceAddressFactorySpiPreferIPV4Test.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+        // shutdown the gateway
+        try {
+            gateway.stop();
+        } catch (Exception e) {
+            assertFalse(e instanceof java.lang.NullPointerException);
+        }
+
+        // cleanup the IPV4 flag
+        System.setProperty(JAVA_NET_PREFER_IPV4_STACK, "false");
```
Comment (dpwspoon) [inline server/src/test/java/org/kaazing/gateway/server/preferedipv4/ResourceAddressFactorySpiPreferIPV4Test.java]: If I was doing this I would put this in an @after test block so it is always ran

File: server/src/test/java/org/kaazing/gateway/server/preferedipv4/ResourceAddressFactorySpiPreferIPV4Test.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        // startup the gateway
+        GatewayConfiguration gc = getGatewayConfig();
+        Gateway gateway = new Gateway();
+        startupTheGatewayWithAssertions(gc, gateway);
+
+        // shutdown the gateway
+        try {
+            gateway.stop();
+        } catch (Exception e) {
+            assertFalse(e instanceof java.lang.NullPointerException);
```
Comment (dpwspoon) [inline server/src/test/java/org/kaazing/gateway/server/preferedipv4/ResourceAddressFactorySpiPreferIPV4Test.java]: Consider using an expected exception for junit: https://github.com/junit-team/junit/wiki/Exception-testing

File: server/src/test/java/org/kaazing/gateway/server/preferedipv4/ResourceAddressFactorySpiPreferIPV4Test.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    /**
+    * Helper method returning a gateway config
+    * 
+    * @return
+    */
+    private GatewayConfiguration getGatewayConfig() {
+        GatewayConfiguration gc = new GatewayConfigurationBuilder().
+                service().
```
Comment (dpwspoon) [inline server/src/test/java/org/kaazing/gateway/server/preferedipv4/ResourceAddressFactorySpiPreferIPV4Test.java]: Consider turning [off the auto formatter](http://stackoverflow.com/questions/1820908/how-to-turn-off-the-eclipse-code-formatter-for-certain-sections-of-java-code) around the configuration block

File: server/src/test/java/org/kaazing/gateway/server/preferedipv4/ResourceAddressFactorySpiPreferIPV4Test.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+/**
+ * Class for testing preferIPv4Stack behavior
+ */
+public class ResourceAddressFactorySpiPreferIPV4Test {
+
+    private static final String JAVA_NET_PREFER_IPV4_STACK = "java.net.preferIPv4Stack";
+
+    @Test
+    public void testEmptyCollectionWithFlagTrue() {
```
Comment (dpwspoon) [inline server/src/test/java/org/kaazing/gateway/server/preferedipv4/ResourceAddressFactorySpiPreferIPV4Test.java]: No need for test prefix in junit for the methods, as they are identified by class name#test so ResourceAddressFactorySpiPreferIPV4Test#testEmptyCollectionWithFlagTrue is redundant.  Naming is hard, I would suggest ResourceAddressFactorySpiPreferIPV4Test#failsStartupWhenMatchingIPv6AddressesNotFound, though you can probably come up with something better

**Comment by dpwspoon** [comment]
@vmaraloiu, this is all very good.  Apologies, I must of missed this PR yesterday.  Consider the changes above, and ping me when you want me to look at it again.  Thanks!

**Comment by vmaraloiu** [comment]
@dpwspoon : Thank you David for previous reviews. I have checked in a new set of changes, can you please review them?

**Comment by dpwspoon** [comment]
LGTM.  Thanks

Note: our check style plugin has fairly strict conventions for JavaDoc conventions.  All first lines must end in periods and things like that.

**Pull request merged**